sled-agent/src/support_bundle/logs.rs

```diff
@@ -0,0 +1,253 @@
+        let valid_dirs =
+            tokio::task::spawn_blocking(|| sled_diagnostics::valid_log_dirs())
+                .await
+                .map_err(Error::Join)?
+                .map_err(Error::Logs)?;
+
+        let normalized = normalize_path(req);
+        let valid_path = validate_log_dir(&valid_dirs, normalized)?;
+        let mut file = match tokio::fs::File::open(&valid_path).await {
+            Ok(file) => file,
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                return Err(Error::NotFound(valid_path));
+            }
+            Err(e) => return Err(e.into()),
+        };
+        let metadata = file.metadata().await?;
+
+        if !metadata.is_file() {
+            return Err(Error::NotAFile);
+        }
+
+        const CONTENT_TYPE: http::HeaderValue =
+            http::HeaderValue::from_static("text/plain");
+        let content_type = Some(CONTENT_TYPE);
+
+        if let Some(range) = range {
+            // If this has a range request, we need to validate the range
+            // and put bounds on the part of the file we're reading.
+            let range = match range.parse(metadata.len()) {
+                Ok(range) => range,
+                Err(err_response) => return Ok(err_response),
+            };
+
+            info!(&self.log, "SupportBundle GET log file (ranged)";
+                "file" => %valid_path,
+                "start" => range.start(),
+                "limit" => range.content_length().get(),
+            );
+
+            file.seek(std::io::SeekFrom::Start(range.start())).await?;
+            let limit = range.content_length().get();
+            return Ok(range_requests::make_get_response(
+                Some(range),
+                metadata.len(),
+                content_type,
+                ReaderStream::new(file.take(limit)),
+            )?);
+        } else {
+            info!(&self.log, "SupportBundle GET log file";
+                "file" => %valid_path,
+            );
+
+            Ok(range_requests::make_get_response(
+                None,
+                metadata.len(),
+                content_type,
+                // The take here is load bearing as we want to cap continously
+                // growing log files.
+                ReaderStream::new(file.take(metadata.len())),
+            )?)
+        }
+    }
+}
+
+/// Given a path represented as a `str` return a normalized `Utf8PathBuf` that
+/// starts with "/" and removes "..", "."  components.
+fn normalize_path(path: &str) -> Utf8PathBuf {
+    let path = Utf8PathBuf::from(format!("/{path}"));
+    let mut normalized = Utf8PathBuf::new();
+
+    for component in path.components() {
+        match component {
+            camino::Utf8Component::RootDir => normalized.push("/"),
+            camino::Utf8Component::Normal(c) => normalized.push(c),
+            // Toss the rest out rather than popping off the stack. Consumers
+            // should be passing full paths.
```

Comment:
It seems very surprising to me to silently "drop" `..` and `.` components in a path, rather than rejecting them. Should we be returning a "Bad args" error in this case?

Reply:
Maybe sanitize should have been used instead of normalize.
We could certainly reject any request that contains those path components rather than removing them, should we just do that instead?